plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoPageSource.java

```diff
@@ -303,33 +302,33 @@ else if (isMapType(type)) {
             if (value instanceof Map<?, ?> document) {
                 BlockBuilder builder = output.beginBlockEntry();
                 for (Map.Entry<?, ?> entry : document.entrySet()) {
-                    appendTo(type.getTypeParameters().get(0), entry.getKey(), builder);
-                    appendTo(type.getTypeParameters().get(1), entry.getValue(), builder);
+                    appendTo(mapType.getKeyType(), entry.getKey(), builder);
+                    appendTo(mapType.getValueType(), entry.getValue(), builder);
                 }
                 output.closeEntry();
                 return;
             }
         }
-        else if (isRowType(type)) {
+        else if (type instanceof RowType rowType) {
             if (value instanceof Map<?, ?> mapValue) {
                 BlockBuilder builder = output.beginBlockEntry();
 
                 List<String> fieldNames = new ArrayList<>();
-                for (int i = 0; i < type.getTypeSignature().getParameters().size(); i++) {
-                    TypeSignatureParameter parameter = type.getTypeSignature().getParameters().get(i);
-                    fieldNames.add(parameter.getNamedTypeSignature().getName().orElse("field" + i));
+                for (int i = 0; i < rowType.getFields().size(); i++) {
+                    RowType.Field field = rowType.getFields().get(i);
+                    fieldNames.add(field.getName().orElse("field" + i));
                 }
-                checkState(fieldNames.size() == type.getTypeParameters().size(), "fieldName doesn't match with type size : %s", type);
-                for (int index = 0; index < type.getTypeParameters().size(); index++) {
-                    appendTo(type.getTypeParameters().get(index), mapValue.get(fieldNames.get(index)), builder);
+                checkState(fieldNames.size() == rowType.getFields().size(), "fieldName doesn't match with type size : %s", type);
+                for (int index = 0; index < rowType.getFields().size(); index++) {
+                    appendTo(rowType.getTypeParameters().get(index), mapValue.get(fieldNames.get(index)), builder);
                 }
                 output.closeEntry();
                 return;
             }
             if (value instanceof DBRef dbRefValue) {
                 BlockBuilder builder = output.beginBlockEntry();
 
-                checkState(type.getTypeParameters().size() == 3, "DBRef should have 3 fields : %s", type);
+                checkState(rowType.getFields().size() == 3, "DBRef should have 3 fields : %s", type);
                 appendTo(type.getTypeParameters().get(0), dbRefValue.getDatabaseName(), builder);
```

Comment:
Instead of `getTypeParameters` - can we get the type from the field ?

Reply:
Updated.